### DISCUSSION TITLE
Symfony - Create a Form Type with no collection name

### DISCUSSION THREAD

**Discussion by bhagman** [symfony/symfony#49854] [Q&A] [answered]
I'm trying to create a controller which accepts parameters via the query string, and uses a Symfony Form Type.  I require the query string to be as clean as possible (e.g. `?var1=value1&var2=value2`).

How can I create a Form Type that has a no prefix in the form parameter names?

If you create a Form in Symfony, the default method is 'POST', and the parameters are all pseudo-named as a collection the form name.

If I wanted to use the 'GET' method, for instance:
```php
class SearchFormType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options): void
    {
        $builder
            ->setMethod('GET')
            ->add('searchText', TextType::class);
    }
}
```

When created (and eventually rendered) in a controller:

```php
        $form = $this->createForm(SearchFormType::class);
```

This results in a URL with the following query string: `?form%5BsearchText%5D=something`

One [suggestion](https://stackoverflow.com/questions/13384056/symfony2-1-using-form-with-method-get/13474522) is to use `createNamed()`/`createNamedBuilder()` from the Form Builder, but this is implied to be used within the controller context.

What is the best way to have a Form Type (i.e. a FormType class) with no form collection name, resulting in a query string `?searchText=something`?

**Comment by norkunas** [comment]
Alternatively you can add `getBlockPrefix() { return ''; }` to your form type

**Reply by bhagman** [reply]
I tried this after you suggested it and it works.

I have seen this idea referenced on some SO [questions](https://stackoverflow.com/questions/49348450/), however it is suggested this isn't the best solution.  I'm unclear as to why it is or isn't a good solution.

It feels to me that this should be a common problem (i.e. clean URLs with uncomplicated query string parameters), and that there is no obvious solution in Symfony Forms.  Am I wrong in my thinking or approach?  For example, should I not be using a full Form Type for a "search" field, and instead just build and handle the form in the controller?

**Reply by norkunas** [reply]
If you have just one field for your case you can totally do that without forms, just fetch the value from the query string :)
Yes, the empty block prefix solution is more of a hack, but if you don't need to customize it using block prefixes then imho it's enough
or then use a proper solution with `formFactory->createNamed('')`
> but this is implied to be used within the controller context.

do you want to handle the form somewhere else?

**Reply by bhagman** [reply]
Perhaps I'm expecting to be able to set the form name/prefix using a parameter through the `createForm()` function in the controller, or to be able to set it through the configuration options in the Form Type itself, instead of "hacking" it away with a `getBlockPrefix()` overload.

The solution by @ker0x is the closest I'll get to the "proper" way to do it.

I must be approaching this wrong, or not many people need this solution.

Thanks for your help!

**Comment by ker0x** [comment, marked as answer]
In a controller, you can call the form factory directly:

```php
$form = $this->container->get('form.factory')->createNamed('', SearchFormType::class);
```

Otherwise, in a service you have to inject it:

```php
class MyService 
{
    public function __construct(private FormFactoryInterface $formFactory)
    {
    }

    public function someMethod()
    {
        $form = $this->formFactory->createNamed('', SearchFormType::class);
    }
}
```

**Comment by bhagman** [comment]
Thanks -- although slightly unintuitive, this is what I was looking for.

Being able to set the form collection during form creation makes sense.  However, having to dance around to get the FormFactory looks pretty ugly - i.e. no beginner would think of getting the "form.factory" from the controller's container and create a form from that.  It seems to me that there should be a better way to control the form name either as a configuration option in the Form Type's `configureOptions()`, or directly from the controller but with a more intuitive function call (e.g. `$form = $this->createNamedForm('', SearchFormType::class);`)

I still suspect I'm approaching this incorrectly, or it's not a common problem.

Thanks for your help!

**Reply by ker0x** [reply]
It's not a common use case indeed but it's already explain in the documentation: [Changing the Form Name](https://symfony.com/doc/current/forms.html#changing-the-form-name)

**Reply by ker0x** [reply]
And if you need it a lot in your controllers, you can always create a trait with it:

```php
/**
  * Creates and returns a named form instance from the type of the form.
  */
protected function createFormNamed(string $name, string $type, mixed $data = null, array $options = []): FormInterface
{
    return $this->container->get('form.factory')->createNamed($name, $type, $data, $options);
}
```
